tests/unit/test_utils.py

```diff
@@ -0,0 +1,102 @@
+
+    # parses dates in dicts in lists
+    expected_datetime = datetime(2016, 11, 14, 11, 10, 52, 425000, timezone.utc)
+    assert _parse_date_fields([{'createdAt': '2016-11-14T11:10:52.425Z'}]) == [{'createdAt': expected_datetime}]
+
+    # parses nested dates
+    expected_datetime = datetime(2020, 2, 29, 10, 9, 8, 100000, timezone.utc)
+    assert _parse_date_fields({'a': {'b': {'c': {'createdAt': '2020-02-29T10:09:08.100Z'}}}}) == {'a': {'b': {'c': {'createdAt': expected_datetime}}}}
+
+    # doesn't parse dates nested too deep
+    expected_datetime = datetime(2020, 2, 29, 10, 9, 8, 100000, timezone.utc)
+    assert _parse_date_fields({'a': {'b': {'c': {'d': {'createdAt': '2020-02-29T10:09:08.100Z'}}}}}) \
+        == {'a': {'b': {'c': {'d': {'createdAt': '2020-02-29T10:09:08.100Z'}}}}}
+
+    # doesn't die when the date can't be parsed
+    assert _parse_date_fields({'createdAt': 'NOT_A_DATE'}) == {'createdAt': 'NOT_A_DATE'}
+
+
+def test__pluck_data() -> None:
+    # works correctly when data is present
+    assert _pluck_data({'data': {}}) == {}
+    assert _pluck_data({'a': 'b', 'data': {'b': 'c'}}) == {'b': 'c'}
+
+    # throws the right error when it is not
+    with pytest.raises(ValueError):
+        _pluck_data({'a': 'b'})
+    with pytest.raises(ValueError):
+        _pluck_data(None)
+    with pytest.raises(ValueError):
+        _pluck_data('{"a": "b"}')
+
+
+def test__retry_with_exp_backoff() -> None:
+    attempt_counter = 0
+
+    class RetryableException(Exception):
+        pass
+
+    class BailException(Exception):
+        pass
+
+    def returns_on_fifth_attempt(bail: Callable, attempt: int) -> Any:
+        nonlocal attempt_counter
+        attempt_counter += 1
+
+        if attempt == 5:
+            return 'SUCCESS'
+        raise RetryableException()
+
+    def bails_on_third_attempt(bail: Callable, attempt: int) -> Any:
+        nonlocal attempt_counter
+        attempt_counter += 1
+
+        if attempt == 3:
+            bail(BailException())
+        raise RetryableException()
+
+    # Returns the correct result after the correct time (should take 100 + 200 + 400 + 800 = 1500 ms)
+    start = time.time()
+    result = _retry_with_exp_backoff(returns_on_fifth_attempt, backoff_base_millis=100, backoff_factor=2, random_factor=0)
+    elapsed_time_seconds = time.time() - start
+    assert result == 'SUCCESS'
+    assert attempt_counter == 5
+    assert elapsed_time_seconds > 1.4
+    assert elapsed_time_seconds < 2.0
+
+    # Stops retrying when failed for max_retries times
+    attempt_counter = 0
+    with pytest.raises(RetryableException):
+        _retry_with_exp_backoff(returns_on_fifth_attempt, max_retries=4, backoff_base_millis=1)
+    assert attempt_counter == 4
+
+    # Bails when the bail function is called
+    attempt_counter = 0
+    with pytest.raises(BailException):
+        _retry_with_exp_backoff(bails_on_third_attempt, backoff_base_millis=1)
+    assert attempt_counter == 3
```

Comment:
Let's use native python `unittest` framework for this. When you create a class with unit tests which subclasses `unittest.TestCase` you get lot of fancy comparison methods such as `self.assertEqual()` that ar easier to read than plain `assert X == Y`

```
import time
import unittest
from datetime import datetime, timezone
from typing import Any, Callable

from apify_client._utils import _parse_date_fields, _pluck_data, _retry_with_exp_backoff, _to_safe_id


class UtilsTest(unittest.TestCase):

    def test__to_safe_id(self) -> None:
        self.assertEqual(_to_safe_id('abc'), 'abc')
        self.assertEqual(_to_safe_id('abc/def'), 'abc~def')
        self.assertEqual(_to_safe_id('abc/def'), 'abc~def')

    def test__parse_date_fields(self) -> None:
        # works correctly on empty dicts
        self.assertEqual(_parse_date_fields({}), {})

        # correctly parses dates on fields ending with -At
        expected_datetime = datetime(2016, 11, 14, 11, 10, 52, 425000, timezone.utc)
        self.assertEqual(
            _parse_date_fields({'createdAt': '2016-11-14T11:10:52.425Z'}),
            {'createdAt': expected_datetime}
        )

        # doesn't parse dates on fields not ending with -At
        self.assertEqual(
            _parse_date_fields({'saveUntil': '2016-11-14T11:10:52.425Z'}),
            {'saveUntil': '2016-11-14T11:10:52.425Z'}
        )

        # parses dates in dicts in lists
        expected_datetime = datetime(2016, 11, 14, 11, 10, 52, 425000, timezone.utc)
        self.assertEqual(
            _parse_date_fields([{'createdAt': '2016-11-14T11:10:52.425Z'}]),
            [{'createdAt': expected_datetime}]
        )

        # parses nested dates
        expected_datetime = datetime(2020, 2, 29, 10, 9, 8, 100000, timezone.utc)
        self.assertEqual(
            _parse_date_fields({'a': {'b': {'c': {'createdAt': '2020-02-29T10:09:08.100Z'}}}}),
            {'a': {'b': {'c': {'createdAt': expected_datetime}}}}
        )

        # doesn't parse dates nested too deep
        expected_datetime = datetime(2020, 2, 29, 10, 9, 8, 100000, timezone.utc)
        self.assertEqual(
            _parse_date_fields({'a': {'b': {'c': {'d': {'createdAt': '2020-02-29T10:09:08.100Z'}}}}}),
            {'a': {'b': {'c': {'d': {'createdAt': '2020-02-29T10:09:08.100Z'}}}}}
        )

        # doesn't die when the date can't be parsed
        self.assertEqual(
            _parse_date_fields({'createdAt': 'NOT_A_DATE'}),
            {'createdAt': 'NOT_A_DATE'}
        )

    def test__pluck_data(self) -> None:
        # works correctly when data is present
        self.assertEqual(_pluck_data({'data': {}}), {})
        self.assertEqual(_pluck_data({'a': 'b', 'data': {'b': 'c'}}), {'b': 'c'})

        # throws the right error when it is not
        with self.assertRaises(ValueError):
            _pluck_data({'a': 'b'})
        with self.assertRaises(ValueError):
            _pluck_data(None)
        with self.assertRaises(ValueError):
            _pluck_data('{"a": "b"}')

    def test__retry_with_exp_backoff(self) -> None:
        attempt_counter = 0

        class RetryableException(Exception):
            pass

        class BailException(Exception):
            pass

        def returns_on_fifth_attempt(bail: Callable, attempt: int) -> Any:
            nonlocal attempt_counter
            attempt_counter += 1

            if attempt == 5:
                return 'SUCCESS'
            raise RetryableException()

        def bails_on_third_attempt(bail: Callable, attempt: int) -> Any:
            nonlocal attempt_counter
            attempt_counter += 1

            if attempt == 3:
                bail(BailException())
            raise RetryableException()

        # Returns the correct result after the correct time (should take 100 + 200 + 400 + 800 = 1500 ms)
        start = time.time()
        result = _retry_with_exp_backoff(returns_on_fifth_attempt, backoff_base_millis=100, backoff_factor=2, random_factor=0)
        elapsed_time_seconds = time.time() - start
        self.assertEqual(result, 'SUCCESS')
        self.assertEqual(attempt_counter, 5)
        self.assertGreater(elapsed_time_seconds, 1.4)
        self.assertLess(elapsed_time_seconds, 2.0)

        # Stops retrying when failed for max_retries times
        attempt_counter = 0
        with self.assertRaises(RetryableException):
            _retry_with_exp_backoff(returns_on_fifth_attempt, max_retries=4, backoff_base_millis=1)
        self.assertEqual(attempt_counter, 4)

        # Bails when the bail function is called
        attempt_counter = 0
        with self.assertRaises(BailException):
            _retry_with_exp_backoff(bails_on_third_attempt, backoff_base_millis=1)
        self.assertEqual(attempt_counter, 3)

```

Reply:
To me, as somebody who doesn't know python at all, this syntax

`assert _to_safe_id('abc~def') == 'abc~def'`

is easier to read than

`self.assertEqual(_to_safe_id('abc/def'), 'abc~def')`

but that is not the right metric. But I remember that guys on the call were mentioning pytest as their tool of choice, isn't it?

Reply:
`pytest` is the thing which runs these tests under the hood (they support this paradigm).

The advantage of `self.assertXY` is that it's much more robust than `assert X == Y`. You can do all kinds of asserts, you can expect exceptions etc.

List of asserts is here: https://docs.python.org/3/library/unittest.html#assert-methods
You can even check that lists have same elements regardless of order with `assertCountEqual` (this is super useful when comparing lists that results as taking some keys / values from dictionaries :))

Reply:
Updated it to use `self.assertEqual`, it is indeed better, it displays errors better etc.